Zend/zend_object_handlers.c

```diff
@@ -950,17 +931,27 @@ ZEND_API zval *zend_std_write_property(zend_object *zobj, zend_string *name, zva
 	if (EXPECTED(IS_VALID_PROPERTY_OFFSET(property_offset))) {
 try_again:
 		variable_ptr = OBJ_PROP(zobj, property_offset);
+
+		if (prop_info && UNEXPECTED(prop_info->flags & (ZEND_ACC_READONLY|ZEND_ACC_PPP_SET_MASK))
+		 && (Z_TYPE_P(variable_ptr) != IS_UNDEF || (Z_PROP_FLAG_P(variable_ptr) & IS_PROP_UNINIT) || !zobj->ce->__set)) {
```

Comment:
The `!zobj->ce->__set` is not covered by tests

Reply:
Added a test for both `__set` and `__unset`.